### PR TITLE
#8/Implenent-OAuth

### PR DESCRIPTION
**Pull request by denis-onder**
Issue #8 

TL;DR - What does this implement?
---------------------------------
OAuth

How to Test
-----------
Start the Node server using `npm start`,
then run `npm test`.

Worries / Concerns
------------------
Stability and code quality. I believe due to refactoring that I have wrote a bit of "janky" code, so if there's anything you deem refactoring-worthy, please let me know.

Checklist
---------

 - [X] Do we need to test this?
 - [ ] Were tests written?
 - [X] Does this meet coding standards?
 - [X] Is this a code change?
 - [ ] Is this a meta change (e.g issue template)?
 - [X] Is this pull request ready for review?
 - [ ] Is this a breaking change (e.g not backwards compatible with clients)?
 
ESLint Output
-------------
```
Warning: React version was set to "detect" in eslint-plugin-react settings, but the "react" package is not installed. Assuming latest React version for linting.
```

Reviewers
---------
@gmemstr

### CODE REVIEW

File: app/auth/config.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+// const facebookStrategy = require('./strategies/facebook');
+
+module.exports = (passport, refresh) => {
+  passport.serializeUser((user, cb) => cb(null, user));
+  passport.deserializeUser((obj, cb) => cb(null, obj));
+  jwtStrategy(passport);
+  googleStrategy(passport, refresh);
+  twitterStrategy(passport);
+  discordStrategy(passport, refresh);
+  // facebookStrategy(passport);
```
Comment (gmemstr) [inline app/auth/config.js]: Should we keep the FB code around?

File: app/auth/strategies/discord.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
+/* eslint-disable no-param-reassign */
+const { Strategy } = require('passport-discord');
+const { clientID, clientSecret } = require('../../config').oauth.discord;
+const User = require('../../db/models/user.model');
+const registerUser = require('../../functions/registerUser');
+const generateRandomBytes = require('../../functions/generateRandomBytes');
+
+module.exports = (passport, refresh) => {
```
Comment (gmemstr) [inline app/auth/strategies/discord.js]: How much of these different strategies can we consolidate into generic functions?

File: app/auth/strategies/twitter.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      callbackURL: '/auth/twitter/callback',
+    },
+    async (token, tokenSecret, profile, done) => {
+      try {
+        const user = await User.findOne({ where: { clientID: profile.id } });
+        if (user) {
+          return done(null, user);
+        }
+        registerUser(
+          profile.id, // FIXME: Find a better solution than this
```
Comment (gmemstr) [inline app/auth/strategies/twitter.js]: Let's find a better solution now - What are our options?

File: .env.example [hunk 4]
```diff
@@ -8,4 +8,12 @@ MAIL_PORT=
 MAIL_USER=
 MAIL_PASSWORD=
 MAIL_SENDER=
-SECRET=
\ No newline at end of file
+SECRET=
+GOOGLE_OAUTH_CLIENT_ID=
```
Comment (gmemstr) [inline .env.example]: This .env file is starting to get a bit complex, but I'm not sure what the best solution is. :thinking:

File: app/middleware/checkForUser.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    // Functions
+    const checkIfAnUserExists = () => {
+      if (user) return res.status(403).json({ error: 'A user already exists.' });
+    };
+    const checkIfAnUserDoesNotExist = () => {
+      if (!user) return res.status(404).json({ error: 'User not found.' });
+    };
+    // Conduct check
+    /*
+    If we are hitting an endpoint
```
Comment (gmemstr) [inline app/middleware/checkForUser.js]: Comment blocks are misaligned. Also make sure to follow the style guide - should look like

```javascript
/**
 * This is a multi
 * line comment.
 */
```

https://github.com/airbnb/javascript#comments--multiline

Comment (gmemstr) [reply app/middleware/checkForUser.js]: Also make sure these are the max length of 80 characters, I think these are less than that so we can take up fewer lines.

File: app/router.js [hunk 6]
```diff
@@ -1,43 +1,81 @@
 /*
   List routes in the exported function
   The controller methods handle the logic
-*/
+  */
```
Comment (gmemstr) [inline app/router.js]: Comment blocks

https://github.com/airbnb/javascript#comments--multiline

File: app/router.js [hunk 7]
```diff
@@ -1,43 +1,81 @@
...
+  );
+  // OAuth
+  app.get('/auth/google', passport.authenticate('google', opts.google));
+  app.get('/auth/google/callback', passport.authenticate('google', opts.callback));
+  app.get('/auth/twitter', passport.authenticate('twitter'));
+  app.get('/auth/twitter/callback', passport.authenticate('twitter', opts.callback));
+  app.get('/auth/discord', passport.authenticate('discord'));
+  app.get('/auth/discord/callback', passport.authenticate('discord', opts.callback));
+  // app.get('/auth/facebook', passport.authenticate('facebook'));
+  // app.get('/auth/facebook/callback', passport.authenticate('facebook', opts.callback));
```
Comment (gmemstr) [inline app/router.js]: Are we keeping Facebook?

File: app/auth/strategies/twitter.js (hunk 3 above)
Comment (denis-onder) [reply app/auth/strategies/twitter.js]: Since the Twitter API doesn't return an email, we are kinda screwed. We could McGyver a ridiculous string, but I don't think that would be plausible. I'd rather replace the Twitter strategy with Facebook or ditch 'em both if we are still not going to use it.

File: app/auth/config.js (hunk 1 above)
Comment (denis-onder) [reply app/auth/config.js]: A subject to discuss with everyone else, because of the twitter strategy issue.

File: .env.example (hunk 4 above)
Comment (denis-onder) [reply .env.example]: Personally, no idea. It is getting ridiculous and I can't think of anything else to be honest.

File: app/middleware/checkForUser.js (hunk 5 above)
Comment (denis-onder) [reply app/middleware/checkForUser.js]: Roger!

File: app/router.js (hunk 7 above)
Comment (denis-onder) [reply app/router.js]: A subject to discuss with everyone else, because of the twitter strategy issue.

File: app/router.js (hunk 6 above)
Comment (denis-onder) [reply app/router.js]: Will get fixed asap

File: app/auth/strategies/discord.js (hunk 2 above)
Comment (denis-onder) [reply app/auth/strategies/discord.js]: I thought that this was a good idea for the sake of modularity. I know that they are going against DRY, but I can't think of a better way to write them right now. Any ideas?

Comment (denis-onder) [reply app/auth/strategies/discord.js]: Refactored this bit. Ready for another review
